api_tests/harness.js

```diff
@@ -16,202 +25,274 @@ const project_root = execSync("git rev-parse --show-toplevel", {
+        for (let [name, btsieve_config] of btsieves) {
+            if (this.running_btsieves[name]) {
+                continue;
             }
 
-            if (config.comit_node) {
-                Object.keys(config.comit_node).forEach(async function(name) {
-                    const subprocess = await startComitNode(
-                        name,
-                        config.comit_node[name]
-                    );
-                    subprocesses.push(subprocess);
-                });
+            this.running_btsieves[name] = await spawn(
+                project_root + "/target/debug/btsieve",
+                [],
+                {
+                    cwd: test_root,
+                    encoding: "utf-8",
+                    env: btsieve_config.env,
+                    stdio: [
+                        "ignore",
+                        fs.openSync(log_dir + "/btsieve-" + name + ".log", "w"),
+                        fs.openSync(log_dir + "/btsieve-" + name + ".log", "w"),
+                    ],
+                }
+            );
+        }
+    }
+
+    async stopBtsieves() {
+        let names = Object.keys(this.running_btsieves);
+
+        if (names.length > 0) {
+            console.log("Stopping Btsieve(s): " + names.join(", "));
+            for (let process of Object.values(this.running_btsieves)) {
+                process.kill();
             }
+        }
 
-            block_interval = setInterval(() => {
-                generateBlock(config.ledgers, ledgers_config);
-            }, 3000);
+        this.running_btsieves = {};
+    }
+}
 
-            await util.sleep(2000);
-        });
+async function run_tests(test_files) {
+    let ledger_runner = new LedgerRunner();
+    let node_runner = new ComitRunner();
+    let btsieve_runner = new BtsieveRunner();
 
-        require(file);
+    let clean_up = () => {};
 
-        after(async function() {
-            clearInterval(block_interval);
-            this.timeout(ledger_down_time);
-            await cleanUp(subprocesses);
-        });
+    process.once("SIGINT", () => {
+        console.log("SIGINT RECEIEVED");
+        clean_up();
     });
-}
 
-let items = fs.readdirSync(test_dir);
+    clean_up = () => {
+        console.log("cleaning up");
+        btsieve_runner.stopBtsieves();
+        node_runner.stopComitNodes();
+        ledger_runner.stopLedgers();
+        console.log("cleanup done");
+        process.exit();
+    };
+
+    for (let test_file of test_files) {
+        let test_dir = path.dirname(test_file);
+        let config = Toml.parse(
+            fs.readFileSync(test_dir + "/config.toml", "utf8")
+        );
+        global.harness.config = config;
+
+        const mocha = new Mocha({
+            bail: true,
+            ui: "bdd",
+        });
 
-for (let item of items) {
-    if (item.endsWith(".js")) {
-        run_tests(test_dir + "/" + item);
+        for (let item of test_files) {
+            mocha.addFile(test_file);
```

Comment:
❓I'm not sure what is happening here and why, but do you intend to add `test_file` and not `item` ?

Reply:
WOOPS